runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java

```diff
@@ -0,0 +1,139 @@
+import org.corfudb.runtime.collections.SMRMap$CORFUSMR;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.runtime.object.transactions.TransactionType;
+
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Globally Unique Identity (GUID) generator that returns ids
+ * with weak comparable ordering with minimal distributed state sync.
+ * On startup increment a globally unique id via Corfu Object.
+ * Use this id with locally incrementing values to return unique ids.
+ * Higher bits are made of an UTC timestamp while lower bits carry the
+ * counter.
+ *
+ * Created by Sundar Sridharan on 5/22/19.
+ */
+@Slf4j
+public class CorfuGuidGenerator implements OrderedGuidGenerator {
+    private final static int TIMESTAMP_SHIFT            = 24;
+    private final static int INSTANCE_ID_SHIFT          = 8;
+    private final static int MAX_TIMESTAMP_CORRECTION   = 0xFF;
+    private final static int MAX_INSTANCE_ID            = 0xFFff;
+
+    private final String GUID_STREAM_NAME = "CORFU_GUID_COUNTER_STREAM";
+    private final Integer GUID_STREAM_KEY = 0xdeadbeef;
+
+    private final SMRMap<Integer, Long> distributedCounter;
+    private final CorfuRuntime runtime;
+
+    /**
+     * Initialize timestamp & correction to MAX to force an update on the instance id on first call.
+     * This prevents the instance id from bumping up if there are no calls made to this class.
+     */
+    private long timestampCorrection = MAX_TIMESTAMP_CORRECTION;
+    private long previousTimestamp = Long.MAX_VALUE;
+    private long instanceId = 0L;
+
+    public CorfuGuidGenerator(CorfuRuntime rt) {
+        runtime = rt;
+        try {
+            distributedCounter = (SMRMap<Integer, Long>)
+                    (rt.getObjectsView().build()
+                            .setType(SMRMap.class)
+                            .setStreamName(GUID_STREAM_NAME)
+                            .open());
+        } catch (Exception e) {
+            log.error("Failed to initialize CorfuGuidGenerator", e);
+            throw e;
+        }
+    }
+
+    private long updateInstanceId() {
+        long nextInstanceId = 0L;
+        boolean transactionAborted = false;
+        do {
+            try {
+                runtime.getObjectsView().TXBuild()
+                        .type(TransactionType.OPTIMISTIC)
+                        .build()
+                        .begin();
+
+                nextInstanceId = distributedCounter.getOrDefault(GUID_STREAM_KEY, 1L) + 1;
+                distributedCounter.put(GUID_STREAM_KEY, nextInstanceId);
+
+                runtime.getObjectsView().TXEnd();
+                transactionAborted = false;
+            } catch (TransactionAbortedException e) {
+                log.error("Transaction aborted while trying to update GUID counter", e);
+                transactionAborted = true;
+            }
+        } while (transactionAborted);
+
+        instanceId = nextInstanceId;
```

Comment:
Why track the instanceId if you can just access it from the map?

Reply:
just to avoid another lookup and a lock?